packages/next/next-server/lib/router/router.ts

```diff
@@ -256,6 +261,29 @@ export default class Router implements BaseRouter {
       }
 
       window.addEventListener('popstate', this.onPopState)
+
+      // enable custom scroll restoration handling when available
+      // otherwise fallback to browser's default handling
+      if (manualScrollRestoration) {
+        window.history.scrollRestoration = 'manual'
+
+        let scrollDebounceTimeout: undefined | NodeJS.Timeout
+
+        const debouncedScrollSave = () => {
+          if (scrollDebounceTimeout) clearTimeout(scrollDebounceTimeout)
+
+          scrollDebounceTimeout = setTimeout(() => {
+            const { url, as: curAs, options } = history.state
+            this.changeState('replaceState', url, curAs, {
+              ...options,
```

Comment:
This added an object spread helper. Can we use `Object.assign` instead?

Reply:
Updated